src/common/src/util/value_encoding/row_encoding.rs

```diff
@@ -0,0 +1,225 @@
+    }
+}
+
+/// `RowEncoding` holds row-specific information for Column-Aware Encoding
+struct RowEncoding {
+    flag: Flag,
+    offsets: Vec<u8>,
+    buf: Vec<u8>,
+}
+
+impl RowEncoding {
+    fn new() -> Self {
+        RowEncoding {
+            flag: Flag::EMPTY,
+            offsets: vec![],
+            buf: vec![],
+        }
+    }
+
+    fn set_offsets(&mut self, usize_offsets: &[usize], max_offset: usize) {
+        assert!(self.offsets.is_empty());
+        match max_offset {
+            _n @ ..=const { u8::MAX as usize } => {
+                self.flag |= Flag::OFFSET8;
+                usize_offsets
+                    .iter()
+                    .for_each(|m| self.offsets.put_u8(*m as u8));
+            }
+            _n @ ..=const { u16::MAX as usize } => {
+                self.flag |= Flag::OFFSET16;
+                usize_offsets
+                    .iter()
+                    .for_each(|m| self.offsets.put_u16(*m as u16));
+            }
+            _n @ ..=const { u32::MAX as usize } => {
+                self.flag |= Flag::OFFSET32;
+                usize_offsets
+                    .iter()
+                    .for_each(|m| self.offsets.put_u32(*m as u32));
+            }
+            _ => unreachable!("encoding length exceeds u32"),
+        }
+    }
+
+    fn encode(&mut self, datum_refs: impl Iterator<Item = impl ToDatumRef>) {
+        assert!(
+            self.buf.is_empty(),
+            "should not encode one RowEncoding object multiple times."
+        );
+        let mut offset_usize = vec![];
+        for datum in datum_refs {
+            offset_usize.push(self.buf.len());
+            if let Some(v) = datum.to_datum_ref() {
+                serialize_scalar(v, &mut self.buf);
+            }
+        }
+        let max_offset = *offset_usize
+            .last()
+            .expect("should encode at least one column");
+        self.set_offsets(&offset_usize, max_offset);
+    }
+}
+
+/// Column-Aware `Serializer` holds schema related information, and shall be
+/// created again once the schema changes
+pub struct Serializer {
+    encoded_column_ids: Vec<u8>,
+    datum_num: u32,
+}
+
+impl Serializer {
+    pub fn new(column_ids: &[ColumnId]) -> Self {
+        // currently we hard-code ColumnId as i32
+        let mut encoded_column_ids = Vec::with_capacity(column_ids.len() * 4);
+        for id in column_ids {
+            encoded_column_ids.put_i32_le(id.get_id());
+        }
+        let datum_num = column_ids.len() as u32;
+        Self {
+            encoded_column_ids,
+            datum_num,
+        }
+    }
+
+    pub fn serialize_row_column_aware(&self, row: impl Row) -> Vec<u8> {
+        assert_eq!(row.len(), self.datum_num as usize);
+        let mut encoding = RowEncoding::new();
+        encoding.encode(row.iter());
+        self.serialize(encoding)
+    }
+
+    fn serialize(&self, encoding: RowEncoding) -> Vec<u8> {
+        let mut row_bytes = vec![];
```

Comment:
It'll be better to pre-calculate the capacity and reserve it by `with_capacity` when creating the empty vec. :)